samcli/lib/bootstrap/companion_stack/companion_stack_manager.py

```diff
@@ -0,0 +1,259 @@
+            Returns empty list if companion stack does not exist
+        """
+        if not self.does_companion_stack_exist():
+            return []
+        deployed_repos: List[ECRRepo] = self.list_deployed_repos()
+        current_mapping = self._builder.repo_mapping
+
+        unreferenced_repos: List[ECRRepo] = list()
+        for deployed_repo in deployed_repos:
+            for _, current_repo in current_mapping.items():
+                if current_repo.logical_id == deployed_repo.logical_id:
+                    break
+            else:
+                unreferenced_repos.append(deployed_repo)
+        return unreferenced_repos
+
+    def delete_unreferenced_repos(self) -> None:
+        """
+        Blocking call to delete all deployed ECR repos that are unreferenced by a function
+        """
+        repos = self.get_unreferenced_repos()
+        for repo in repos:
+            try:
+                self._ecr_client.delete_repository(repositoryName=repo.physical_id, force=True)
+            except self._ecr_client.exceptions.RepositoryNotFoundException:
+                pass
+
+    def sync_repos(self) -> None:
+        """
+        Blocking call to sync companion stack with the following actions
+        Create/Update companion stack.
+        Deletes unreferenced repos.
+        Deletes companion stack if there isn't any repo left.
+        """
+        exists = self.does_companion_stack_exist()
+        has_repo = bool(self.get_repository_mapping())
+        if exists:
+            self.delete_unreferenced_repos()
+            if has_repo:
+                self.update_companion_stack()
+            else:
+                self.delete_companion_stack()
+        elif not exists and has_repo:
+            self.update_companion_stack()
+
+    def does_companion_stack_exist(self) -> bool:
+        """
+        Does companion stack exist
+
+        Returns
+        -------
+        bool
+            Returns True if companion stack exists
+        """
+        try:
+            self._cfn_client.describe_stacks(StackName=self._companion_stack.stack_name)
+            return True
+        except ClientError:
+            return False
+
+    def get_repository_mapping(self) -> Dict[str, str]:
+        """
+        Get current function to repo mapping
+
+        Returns
+        -------
+        Dict[str, str]
+            Dictionary with key as function logical ID and value as ECR repo URI.
+        """
+        return dict((k, self.get_repo_uri(v)) for (k, v) in self._builder.repo_mapping.items())
+
+    def get_repo_uri(self, repo: ECRRepo) -> str:
+        """
+        Get repo URI for a ECR repo
+
+        Parameters
+        ----------
+        repo: ECRRepo
+
+        Returns
+        -------
+        str
+            ECR repo URI based on account ID and region.
+        """
+        return repo.get_repo_uri(self._account_id, self._region_name)
```

Comment:
Q: why do we need this to determine repo URI? is there no other way to determine it?

Reply:
Generating repo URI locally will enable the comparison between deployed and to-be-deployed repos. The comparison then can prompt the user which repos will be removed.
This is the same as CFN: `!Sub ${AWS::AccountId}.dkr.ecr.${AWS::Region}.${AWS::URLSuffix}/repo_logical_id` but without going through CFN.